contracts/src/protocol/IPTokenStaking.sol

```diff
@@ -320,231 +327,274 @@ contract IPTokenStaking is IIPTokenStaking, Ownable2StepUpgradeable, ReentrancyG
+            delegationId = ++_delegationIdCounter;
+            duration = stakingDurations[stakingPeriod];
         }
+        emit Deposit(delegatorUncmpPubkey, validatorUncmpPubkey, stakeAmount, duration, delegationId, msg.sender, data);
+        // We burn staked tokens
+        payable(address(0)).transfer(stakeAmount);
 
         _refundRemainder(remainder);
 
-        emit Deposit(delegatorUncmpPubkey, delegatorCmpPubkey, validatorCmpPubkey, stakeAmount);
+        return delegationId;
     }
 
-    // TODO: Redelegation also requires unbonding period to be executed. Should we separate storage for this for el?
-    /// @notice Entry point for redelegating the staked token.
-    /// @dev Redelegateion redelegates staked token from src validator to dst validator (x/staking.MsgBeginRedelegate)
-    /// @param p See RedelegateParams
+    /// @notice Entry point for redelegating the stake to another validator.
+    /// @dev For non flexible staking, your staking period will continue as is.
+    /// @dev For locked tokens, this will fail in CL if the validator doesn't support unlocked staking.
+    /// @param delegatorUncmpPubkey Delegator's 65 bytes uncompressed secp256k1 public key.
+    /// @param validatorUncmpSrcPubkey Validator's 65 bytes uncompressed secp256k1 public key.
+    /// @param validatorUncmpDstPubkey Validator's 65 bytes uncompressed secp256k1 public key.
+    /// @param delegationId The delegation ID, 0 for flexible staking.
+    /// @param amount The amount of stake to redelegate.
     function redelegate(
-        RedelegateParams calldata p
+        bytes calldata delegatorUncmpPubkey,
+        bytes calldata validatorUncmpSrcPubkey,
+        bytes calldata validatorUncmpDstPubkey,
+        uint256 delegationId,
+        uint256 amount
     )
         external
-        verifyUncmpPubkeyWithExpectedAddress(p.delegatorUncmpPubkey, msg.sender)
-        verifyExistingValidator(p.validatorCmpSrcPubkey)
-        verifyExistingValidator(p.validatorCmpDstPubkey)
+        payable
+        verifyUncmpPubkeyWithExpectedAddress(delegatorUncmpPubkey, msg.sender)
+        verifyUncmpPubkey(validatorUncmpSrcPubkey)
+        verifyUncmpPubkey(validatorUncmpDstPubkey)
     {
-        (uint256 stakeAmount, ) = roundedStakeAmount(p.amount);
-        bytes memory delegatorCmpPubkey = Secp256k1.compressPublicKey(p.delegatorUncmpPubkey);
-
-        _redelegate(delegatorCmpPubkey, p.validatorCmpSrcPubkey, p.validatorCmpDstPubkey, stakeAmount);
+        if (keccak256(validatorUncmpSrcPubkey) == keccak256(validatorUncmpDstPubkey)) {
+            revert Errors.IPTokenStaking__RedelegatingToSameValidator();
+        }
+        (uint256 stakeAmount, ) = roundedStakeAmount(msg.value);
+        if (stakeAmount < minStakeAmount) {
+            revert Errors.IPTokenStaking__StakeAmountUnderMin();
+        }
+        emit Redelegate(delegatorUncmpPubkey, validatorUncmpSrcPubkey, validatorUncmpDstPubkey, delegationId, amount);
```

Comment:
It would be add comments to explain why we don't need to refund remainder of roundedStakeAmount().

Reply:
If I understand correctly, the remainder will stay in the source validator. If that’s the case, we might explain how do we handle the remainder in the source validator, especially since it’s likely that the remainder will be less than the `minStakeAmount`. we might better add comments to explain why it is expected state/behavior.